### PR TITLE
chore(payment-processor): batch fee usd limit

### PR DESCRIPTION
**Pull request by olivier7delf**
Add the possibility to add/skip the batch fee amount in USD limit using the payment-processor methods.
e.g: activated, a payer will pay no more than $150 batch fees.

Breaking change: the payment-processor's interface changes drastically

RN:
To make PR understandable, it merges into [chore/smart-contract-batch-fee-usd-limit](https://github.com/RequestNetwork/requestNetwork/tree/chore/smart-contract-batch-fee-usd-limit)].
The PR is OK when merging directly on master.

### CODE REVIEW

**Comment by bertux** [comment]
> Add the possibility to limit batch fee in USD using the payment-processor methods.
> 
> Question: within TESTS (private network) currencyManager cannot be used to get a conversion path from a token to USD, is there another solution to test it?

I've shown @olivier7delf how to add support for private network in currencyManager as it's done for testnets.
So tests will be able to use it.

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts
Comment (bertux) [inline packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:27]: why not adding that to the currency package for the supported tokens in network private as it's done for testnets?

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts
Comment (bertux) [inline packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:276]: why add true which is the default value ?

File: packages/smart-contracts/src/lib/artifacts/BatchConversionPayments/index.ts
Comment (bertux) [inline packages/smart-contracts/src/lib/artifacts/BatchConversionPayments/index.ts:20]: can't merge until it's deployed.

**Review by bertux — CHANGES_REQUESTED** [review]
still need changes to make it deployable but can be merge in the smart contract branch if it helps to solve automatic tests of payment-processor.

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts
Comment (olivier7delf) [reply packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:27]: it breaks others' tests

File: packages/smart-contracts/src/lib/artifacts/BatchConversionPayments/index.ts
Comment (olivier7delf) [reply packages/smart-contracts/src/lib/artifacts/BatchConversionPayments/index.ts:20]: Sure, I am trying to get every approval, for both queries. And in the meantime, to have more ETH on the admin address to deploy it before to update once for all

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts
Comment (olivier7delf) [reply packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:276]: done thx

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts
Comment (yomarion) [inline packages/payment-processor/src/payment/batch-conversion-proxy.ts:54]: That's a bold move... I think it should be set to false by default: the risk of paying more gas is less important than the risk of paying more fees.

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts
Comment (yomarion) [inline packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:14]: Commented code

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts
Comment (yomarion) [inline packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:66]: Weird, usually we have 2 decimals for fiat so this looks like 550.00

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts
Comment (yomarion) [inline packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:419]: Is this details with fee limit or not?

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts
Comment (yomarion) [inline packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:438]: This is odd.

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts
Comment (olivier7delf) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts:54]: I made this choice because it should be false only when a batch worth is above 50K $, which doesn't happen that often yet.
But if you think it is too risky, I am fine to set it to false by default.

File: packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts
Comment (olivier7delf) [reply packages/payment-processor/test/payment/any-to-erc20-batch-proxy.test.ts:419]: without fee limit indeed, updated.

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts
Comment (olivier7delf) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts:114]: it is tested if it has a different network in the loop after.
The method is now updated to return an object

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts
Comment (olivier7delf) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts:274]: If it throws, then, it means you can only batch payments that have a conversion path.
It might be right for RF, but for others builders?

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts
Comment (olivier7delf) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts:109]: I would need to discuss it with someone before working on it.
This would be useful when we plan to use different versions of a proxy, which is not yet the case.

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts
Comment (benjlevesque) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts:35]: what bothers you @yomarion? This is exactly how it's supposed to be used

File: packages/smart-contracts/src/lib/artifacts/BatchConversionPayments/index.ts
Comment (benjlevesque) [inline packages/smart-contracts/src/lib/artifacts/BatchConversionPayments/index.ts:53]:
```suggestion
        // Caution: no ERC20Conversion proxy on fuse
```

File: packages/smart-contracts/src/lib/artifacts/BatchConversionPayments/index.ts
Comment (benjlevesque) [inline packages/smart-contracts/src/lib/artifacts/BatchConversionPayments/index.ts:17]: why the change of address, and not a new version? What will happen to payments already made?

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts
Comment (benjlevesque) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts:35]: scratch that, @yomarion , I just realized it's not a test file

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts
Comment (olivier7delf) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts:95]: The network is now an option.
It makes it easier to use the function: it uses the last batch version by default.
And later, we will be able to update the version management easily without breaking things.

File: packages/smart-contracts/src/lib/artifacts/BatchConversionPayments/index.ts
Comment (olivier7delf) [reply packages/smart-contracts/src/lib/artifacts/BatchConversionPayments/index.ts:17]: As discussed, batch proxy addresses are only used by the payment processor.
Keeping the old addresses (and json) is not needed.

File: packages/payment-processor/src/payment/batch-conversion-proxy.ts
Comment (olivier7delf) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts]: I can but is it a bit early?
As I have only this argument now, and we don't if/when we are gonna update batch payments.

Comment (benjlevesque) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts]: imo adopting this style for all methods would allow to have always the same arguments:

pay(request, version, signer, options) 

where options vary depending on the method 

not a blocker, your call

Comment (olivier7delf) [reply packages/payment-processor/src/payment/batch-conversion-proxy.ts]: I am almost fine with your suggestion: pay(request, signer, options, overrides), or maybe also overrides within options?
But some functions look like pay(request, signer, amount, feeAmount, overrides), or even pay(request, signer, paymentSettings, amount, feeAmount, overrides).

I am down to do if it becomes a kind of pay() convention within RN

File: packages/payment-processor/src/payment/index.ts
Comment (yomarion) [inline packages/payment-processor/src/payment/index.ts:332]: Commented code here

File: packages/payment-processor/src/payment/index.ts
Comment (yomarion) [inline packages/payment-processor/src/payment/index.ts:348]: (already discussed) this should be fixed

**Review by yomarion — APPROVED** [review]

File: packages/payment-processor/src/payment/settings.ts
Comment (olivier7delf) [inline packages/payment-processor/src/payment/settings.ts:57]: added here because "isEscrow?: boolean;" is defined in the interface replicated within invoicing repository

**Pull request merged**
